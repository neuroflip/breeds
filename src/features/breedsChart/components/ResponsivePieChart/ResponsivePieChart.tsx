import { PieChart, Pie, Cell, Legend, ResponsiveContainer, Tooltip } from 'recharts';
import { BreedsData } from '../../slices/types';
import { colorPalette } from './colorPalette';

type BreedsPieChartProps  = {
  data: BreedsData[]
}

const getLabel = function(entry: BreedsData) {
  return `${ entry.value }%`;
}

const generatePieCells = (topTenBreeds: BreedsData[]) => {
  return topTenBreeds.map((_entry, index) => (
    <Cell key={`cell-${index}`} fill={colorPalette[index % colorPalette.length]} />
  ))
}

const ResponsivePieChart = ({ data }: BreedsPieChartProps) => {

  return (<>
    <ResponsiveContainer width="100%" height={350}>
      <PieChart width={400} height={400}>
        <Pie data={ data } cx="50%" cy="50%" labelLine={true}
          label={ getLabel } outerRadius={ 80 }
          dataKey="value" nameKey="name">
            { generatePieCells(data) }
        </Pie>
        <Legend />
        <Tooltip />
      </PieChart>
  </ResponsiveContainer>
  </>)
}

export default ResponsivePieChart