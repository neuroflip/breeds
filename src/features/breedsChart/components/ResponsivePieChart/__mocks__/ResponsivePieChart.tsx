const ResponsivePieChart = () => {
  return <>ResponsivePieChart</>
}

export default ResponsivePieChart