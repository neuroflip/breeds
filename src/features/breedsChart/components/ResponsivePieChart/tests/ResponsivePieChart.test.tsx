import { describe, expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import ResponsivePieChart from '../ResponsivePieChart';
import mockData from './breadsData.json' 

vi.mock('recharts', async () => {
  const actual = await vi.importActual<typeof import('recharts')>('recharts');

  return {
    ...actual,
    PieChart: vi.fn(({ children }) => <div data-testid="PieChart">{children}</div>),
    Pie: vi.fn(() => { return <div data-testid="Pie">{ mockData.map((_e, index) => <div key={index} data-testid="Cell" />) } </div> }),
    Legend: vi.fn(() => <div data-testid="Legend" />),
    ResponsiveContainer: vi.fn(({ children }) => <div data-testid="ResponsiveContainer">{children}</div>),
    Tooltip: vi.fn(() => <div data-testid="Tooltip" />)
  }
});

describe('ResponsivePieChart', () => {
  test('renders the chart parts correctly', () => {
    render(<ResponsivePieChart data={ mockData } />);

    expect(screen.getByTestId('ResponsiveContainer')).toBeInTheDocument();
    expect(screen.getByTestId('PieChart')).toBeInTheDocument();
    expect(screen.getByTestId('Pie')).toBeInTheDocument();
    expect(screen.getByTestId('Legend')).toBeInTheDocument();
    expect(screen.getByTestId('Tooltip')).toBeInTheDocument();
    expect(screen.getAllByTestId('Cell').length).toBe(mockData.length);
  });
});