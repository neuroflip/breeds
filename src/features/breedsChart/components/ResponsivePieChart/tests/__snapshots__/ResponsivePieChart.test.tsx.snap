// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResponsivePieChart > renders the chart parts correctly 1`] = `
<div>
  <div
    data-testid="ResponsiveContainer"
  >
    <div
      data-testid="PieChart"
    >
      <div
        data-testid="Pie"
      >
        <div
          data-testid="Cell"
        />
        <div
          data-testid="Cell"
        />
        <div
          data-testid="Cell"
        />
         
      </div>
      <div
        data-testid="Legend"
      />
      <div
        data-testid="Tooltip"
      />
    </div>
  </div>
</div>
`;
